docs/combinatorics.rst

```diff
@@ -1,6 +1,116 @@
 
-=====================================
-Ranking and Unranking Tree Topologies
-=====================================
-TODO
+=============
+Combinatorics
+=============
+tskit uses a combinatorial approach to identify unique topologies of
+rooted, leaf-labelled trees. It provides methods for enumerating
+all possible tree topologies, as well as converting back and forth
+between a tree and its position, or rank, in the enumeration of all
+possible topologies. These methods do not only apply to binary trees;
+rather, they cover general, rooted trees with no unary nodes.
+
+=================================     =====================================
+:meth:`Tree.rank`                     Return the rank of this tree.
+:meth:`Tree.unrank`                   Return a Tree given its rank and
+                                      a number of leaves.
+:func:`tskit.all_tree_topologies`     Return a generator over all
+                                      leaf-labelled trees of n leaves.
+:func:`tskit.all_tree_shapes`         Return a generator over all
+                                      tree shapes of n leaves.
+:func:`tskit.all_tree_labellings`     Return a generator over all
+                                      labellings of the given tree's shape.
+=================================     =====================================
+
+.. _sec_tree_ranks:
+
++++++++++++++++++++++++
+Interpreting Tree Ranks
++++++++++++++++++++++++
+To understand tree ranks we must look at how leaf-labelled tree topologies
+are enumerated. For example, we can use :func:`tskit.all_tree_topologies`
+to generate all possible topologies of three leaves:
+
+.. code-block:: python
+
+    for t in tskit.all_tree_topologies(num_leaves=3):
+        display(SVG(t.draw(node_labels={0: 0, 1: 1, 2: 2}, order="tree")))
+
+.. image:: _static/topology_0_0.svg
+    :width: 24%
+.. image:: _static/topology_1_0.svg
+    :width: 24%
+.. image:: _static/topology_1_1.svg
+    :width: 24%
+.. image:: _static/topology_1_2.svg
+    :width: 24%
+
+In this sequence, there exist two distinct tree shapes and each shape
+can be labelled in at least one unique way. Given that topologies are
+ordered first by their shape and then by their labelling, a tree
+topology can be uniquely identified by
+
+1.
+    The shape of the tree
+2.
+    The labelling of the tree's shape
+
+We can refer to the first tree in the above enumeration as the
+first labelling of the first shape of trees with three leaves, or tree
+:math:`(0, 0)`. The second tree can be identified as the first labelling
+of the second shape, or :math:`(1, 0)`, and so on.
+This pair of indexes for the shape and labelling of a tree is referred
+to as the rank of the tree, and can be computed using the
+:meth:`Tree.rank` method.
+
+.. code-block:: python
+
+    ranks = [t.rank() for t in tskit.all_tree_topologies(num_leaves=3)]
+    print("Ranks of 3-leaf trees:", ranks)
+
+::
+
+   Ranks of 3-leaf trees: [(0, 0), (1, 0), (1, 1), (1, 2)]
```

Comment:
Out of interest, is there a bijection of these tuples to integers? It would be nicer if we could just define the rank as an integer.

Reply:
It would be *super* nice if the rank was a direct index into ``list(all_trees(n))``

Reply:
_This_ has been bugging me for a month. A naive bijection would essentially be:
```
sum(num_labellings(s, n) for s in range(shape_rank)) + label_rank
```

but I have not thought of a way yet without running through all preceding shapes, because depending on the number of symmetries shapes can be labelled in varying numbers of ways.

Reply:
OK, let's not get caught up on it. If we figure out a bijection we can change things later.

Reply:
I dunno - having a 2d index is kinda nicer because it makes it easier to separate tree shape from labeling?

Reply:
How often would you want to just know about the shape? (labelling means nothing on its own)

Reply:
A bijection would give you both - you could move back and forth between the two representations. I can't see anyone being particularly interested in the actual rank values beyond defining the actual tree. If this rank was also its index in ``all_trees()``, then so much... neater.